File: module_login/src/main/java/cn/jack/module_login/mvvm/vm/RegisterViewModel.java
package cn.jack.module_login.mvvm.vm;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import javax.inject.Inject;
import cn.jack.module_login.contract.IRegisterLisenter;
import cn.jack.module_login.di.DaggerViewModelComponent;
import cn.jack.module_login.mvvm.modle.entity.UserInfo;
import cn.jack.module_login.mvvm.modle.repository.RegisterHttpRepository;
import jack.wrapper.base.mvvm.viewModel.BaseViewModel;
import jack.wrapper.base.mvvm.viewModel.liveData.UIChangeLiveData;

/**
 * @创建者 Jack
 * @创建时间 2021/3/24 15:00
 * @描述
 */
public class RegisterViewModel extends BaseViewModel<RegisterHttpRepository> implements IRegisterLisenter {

    @Inject
    RegisterHttpRepository mRegisterHttpRepository;

    //MutableLiveData来实现双向绑定
    public MutableLiveData<String> mPhone  = new MutableLiveData<>();
    public MutableLiveData<String> mPasswd = new MutableLiveData<>();
    public MutableLiveData<String> mPasswdAgain = new MutableLiveData<>();

    public UIChangeLiveData<UserInfo> navigation2LoginA = new UIChangeLiveData<>();

    public RegisterViewModel(@NonNull Application application) {
        super(application);
        DaggerViewModelComponent.builder().build().inject(this);
        mModel = mRegisterHttpRepository;
    }

    public void registerUser() {
        mModel.register(mPhone.getValue(),mPasswd.getValue(),mPasswdAgain.getValue(),this);
    }

    @Override
    public void registerSuccess(UserInfo userInfo) {
        navigation2LoginA.setValue(userInfo);
    }

    @Override
    public void registerFailed(String msg) {
        showToast(msg);
    }

}

File: module_login/src/main/java/cn/jack/module_login/mvvm/vm/RegisterViewModel.kt
package cn.jack.module_login.mvvm.vm

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import cn.jack.module_login.mvvm.modle.entity.UserInfo
import cn.jack.module_login.mvvm.modle.repository.RegisterHttpRepository
import com.jack.lib_wrapper_mvvm.base.viewmodel.BaseWrapperViewModel
import com.jack.lib_wrapper_net.model.EventResult
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import javax.inject.Inject

/**
 * @创建者 Jack
 * @创建时间 2021/3/24 15:00
 * @描述
 */
@HiltViewModel
class RegisterViewModel @Inject constructor(private val repository: RegisterHttpRepository) :
    BaseWrapperViewModel(repository) {

    private val registerUserInfoState =
        MutableStateFlow<EventResult<UserInfo>>(EventResult.OnComplete)

    val registerUserInfo =
        registerUserInfoState.shareIn(viewModelScope, SharingStarted.WhileSubscribed(5000))

    var mPhone = MutableLiveData<String>()
    var mPasswd = MutableLiveData<String>()
    var mPasswdAgain = MutableLiveData<String>()

    fun registerUser(phone: String,
                     passwd: String,
                     passwdAgain: String) {

        //这种方式实现 无法调用    具体原因不清楚
//        viewModelScope.launch {
//            repository.register("13611113310", "123456", "1234567")
//                .onEach {
//                    registerUserInfoState.value = it
//                    println("===> " + it + " " + registerUserInfoState.value)
//                }
//        }

        repository.register(phone, passwd, passwdAgain)
            .onEach {
                registerUserInfoState.value = it
            }.launchIn(viewModelScope)
    }
}